### PR TITLE
Update property generation for typed properties

### PR DESCRIPTION
**Pull request by elythyr**
In response to https://github.com/phpactor/phpactor/issues/856

I admit I'm a bit tired right now and extremely hungry so I wont go in too much details.
The goal was to be able to generate typed properties.
In order to be able to keep the old generation and the new one depending on project I had to add a management of template by version of PHP.
I also used the occasion to fix one or two thing in the generation of properties.

I made a lot of commit and I try to rely detail each one of them to explain why and how as much as possible.

There is just one thing, a commit that move a blank line from the `ClassLikeUpdater` class to the `Property.php.twig` template create some "problems" in two situations.
I'm open to discussion about that, it would be nice to find a solution but I think we would have to choose the smaller of two issues :)

- [ ] Add tests for PHP7.4 code generation

### CODE REVIEW

File: lib/Adapter/TolerantParser/Updater/ClassLikeUpdater.php [hunk 1]
```diff
@@ -54,41 +55,48 @@ protected function updateProperties(Edits $edits, ClassLikePrototype $classProto
             return;
         }
 
-        $lastProperty = $classMembers->openBrace;
+        $previousMember = $classMembers->openBrace;
```
Comment (dantleech) [inline lib/Adapter/TolerantParser/Updater/ClassLikeUpdater.php]: can we add regression tests for this?

Comment (dantleech) [reply lib/Adapter/TolerantParser/Updater/ClassLikeUpdater.php]: or is this just presentation? if it's not a functional change then the test is not so important.

File: lib/Util/TextFormat.php [hunk 2]
```diff
@@ -18,6 +18,10 @@ public function indent(string $string, int $level = 0)
     {
         $lines = explode(PHP_EOL, $string);
         $lines = array_map(function ($line) use ($level) {
```
Comment (dantleech) [inline lib/Util/TextFormat.php]: test is missing for this class, but it might be good to add

File: tests/Adapter/UpdaterTestCase.php [hunk 3]
```diff
@@ -925,6 +925,7 @@ public function crawl()
                 <<<'EOT'
 class Aardvark
 {
+
     /**
```
Comment (dantleech) [inline tests/Adapter/UpdaterTestCase.php]: hmm, actually this change is controversial for me and is not the SF coding standard or PSR-12, would rather have no space here.

File: tests/Adapter/GeneratorTestCase.php [hunk 4]
```diff
@@ -584,6 +587,7 @@ class Rabbits implements Animal
     const LEGS = 4;
     const SKIN = 'soft';
 
+
```
Comment (dantleech) [inline tests/Adapter/GeneratorTestCase.php]: there should be one blank line between member "types"

**Comment by dantleech** [comment]
Thanks for this and the nice atomic commits:

- :heavy_check_mark:  Always add property after traits & constants

- :heavy_check_mark:OK Move the new line before property with phpdoc in the template
- :heavy_check_mark: Initialize `nextMember` right away
- :heavy_check_mark: Rename `lastProperty` into `previousMember`
- :heavy_check_mark: Fix `TextFormat::indent()` to not indent empty lines

- :x: Prepend a blank line when adding properties for the first time: _looking at the adjusted tests, this isn't compatible with PSR-12, there should be no empty line between the opening brace and the first member declaration_
- :x: Append a blank line after the properties: _there should be only one empty line between member "blocks"_

In general we should be as standard as possible with the coding format.

File: templates/70400/Property.php.twig [hunk 5]
```diff
@@ -0,0 +1,5 @@
+{{ prototype.visibility }}
+{%- if prototype.type.notNone %}
+ {{ prototype.type }} 
+{%- endif %}
+ ${{ prototype.name}}{% if prototype.defaultValue.notNone %} = {{ prototype.defaultValue.export }}{% endif %};
```
Comment (dantleech) [inline templates/70400/Property.php.twig]: I guess we can discuss further in the other PR: but would prefer if we used the standard version notation, e.g. `7.1`. We can then use `version_compare`, and `phpversion()` in the iterator.

Comment (dantleech) [reply templates/70400/Property.php.twig]: Additionally I wondered if we should namespace this, so `templates/php/<current templates>` `templates/php/7.1/<php 7.1 overrides>` etc. But as this space isn't used for anything else I guess it's fine.

We could also make the selection logic part of this library...

**Comment by dantleech** [comment]
I think it would make sense to include the version-specific selection _logic_ in this library rather than the extension (this is where the templates are, and it follows that this library should support them).

The twig-independent logic (for sorting filtering the template paths) could be placed in `Domain`. It can then be integrated in the `TwigRenderer` - which will create a Twig env. if one is not passed (note that in the extension we DO pass a Twig extension).

the nice thing is then that we can add an integration test here for the selection functionality, as it will support version selection out-of-the-box.

File: lib/Adapter/TolerantParser/Updater/ClassLikeUpdater.php (hunk 1 above)
Comment (elythyr) [reply lib/Adapter/TolerantParser/Updater/ClassLikeUpdater.php]: The name it's just presentation.
But the fact that the blank line before new properties was moved to the template it's functional.
Didn't I update the tests already ?

File: tests/Adapter/UpdaterTestCase.php (hunk 3 above)
Comment (elythyr) [reply tests/Adapter/UpdaterTestCase.php]: I would love to too :)
The thing is that I trade a "bug" for another one.
Before the changes if you had
```php
<?php

class Foo
{
    use SomeTrait;

    public function __construct($a)
    {
    }
}
```

And you try to complete the constructor, the property would have been added above the trait declaration, which is not PSR-12 compliant neither.

The problem here is that we want to add a blank line only if there is a phpdoc block for the new property.
But the phpdoc part is totally up to the template, so the updater can't know when to add one or not.
And the templates can't know what is the preview member to decide if the blank line should be avoid or not.

File: tests/Adapter/GeneratorTestCase.php (hunk 4 above)
Comment (elythyr) [reply tests/Adapter/GeneratorTestCase.php]: That's the same issue than above.
I forgot to add tests for the 7.4 version, once done you will see that the problem exists only with the phpdoc.

**Comment by elythyr** [comment]
> Append a blank line after the properties: there should be only one empty line between member "blocks"

I thought that was handled already, I will have to check.

> Also I would prefer we use the version functions, e.g. phpversion versioncompare

I think it's harder to read than simply compare an integer, but whatever you prefer I don't mind.

**Comment by dantleech** [comment]
> I think it's harder to read than simply compare an integer, but whatever you prefer I don't mind.

Harder to read the code maybe -- but in the long term this is exposed to users, so clearer to see in the config `php_version: 7.1` f.e. and the directory names would look more intutive to me.

File: tests/Adapter/UpdaterTestCase.php (hunk 3 above)
Comment (dantleech) [reply tests/Adapter/UpdaterTestCase.php]: hmm, only thing I can suggest is making it as consistent as possible, if we trade one bug for another then let's not add this change.

The problem in general with this library is that it explicitly couples to the presenatation - one solution might be to write a simple CS/whitespace fixer using the Tolerant Parser AST, which can operate on a range of lines.

We could then care less about whitespace concerns when building the source code.

Comment (elythyr) [reply tests/Adapter/UpdaterTestCase.php]: I think it's consistent in the way that it's only in two scenario and only if using phpdoc.
Plus it's 100% compliant and bug free for PHP 7.4.
So I would said that the trade is favorable, otherwise we go back with the previous problems in all versions.

And most important it only appends (if I remember correctly), when you have no properties yet.
So it's in a pretty specific and rare case.

The use of the parser would make it clean of course, if we can manage to check the number of black lines between a declaration member and a function member.
I'm not sure how though, will have to check.

Comment (dantleech) [reply tests/Adapter/UpdaterTestCase.php]: hmm, ok. if the pros outweigh the cons, the let's go with it.

> The user of the parser would ...

I'll try and have a look at this

File: lib/Util/TextFormat.php (hunk 2 above)
Comment (elythyr) [reply lib/Util/TextFormat.php]: Done

**Comment by elythyr** [comment]
The CI is failing because it does not have the modification of `worse-reflection`, once https://github.com/phpactor/worse-reflection/pull/65 will be merged we will need to restart it.

**Comment by elythyr** [comment]
Ok, I have:
- upadted the name of the PR
- remove two commits related to templates override
- I think I took care of every review you made, but I let you make sure of that

So we should be ready to merge once all review comments are closed.
Then the CI of https://github.com/phpactor/code-transform/pull/30 should be OK

**Comment by dantleech** [comment]
@elythyr the "style fixing" PR has been merged now (again) :tada:  so there should be no more problems (:crossed_fingers: ) with whitespace etc. Can we the changes related to that from the PR and update the tests accordingly?

**Comment by dantleech** [comment]
I'll cherry pick the null stuff from this PR ... :)

**Comment by dantleech** [comment]
Nullable types is now supported here: https://github.com/phpactor/code-builder/pull/19 (cherry picked the tests from this PR)

Closing this now - as anything else (e.g. property ordering) in this PR will need to be in a new PR anyway). Thanks for this.

**Comment by elythyr** [comment]
Thanks for that, again I'm sorry about not being active for a while now.
New work, new home, etc. I need to figure out how to reoder my time.
I love this project, I used it daily and I really want to help it getting better & bigger :)
